scripts/docker-build

```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+# To avoid rebuilds we first install all Cargo dependencies
+# This process requires running this script
+set -euo pipefail
+
+# Create a shadow setup under the directory docker/
+# This has only what is needed to install all dependencies
+components=($(find . -type f -name 'Cargo.toml' -exec dirname {} \; | grep -v .docker | sort -u))
+for component in "${components[@]}"; do
+    mkdir -p ".docker/${component}/src"
+    lib_rs="${component}/src/lib.rs"
+    if ! [[ -e ".docker/$lib_rs" ]] ; then
+        touch ".docker/$lib_rs"
+    fi
+    cargo="${component}/Cargo.toml"
+    # Use chksum to avoid unnecessary updates that would ruin the docker cache
+    if [[ ! -e ".docker/$cargo" ]] || [[ $(cksum "$cargo") != $(cksum ".docker/$cargo") ]] ; then
+        cp "$cargo" ".docker/$cargo"
+    fi
+done
```

Comment:
Is it possible to run this script in another builder container and copy the cargo files to the current builder? This avoids failure when user call `docker build` directly

Reply:
Effectively we are using a standard technique for a single Cargo file Rust project. In that case one does not need pre-scripting. The real reason we have a script is to *automatically* find all Cargo files. So if someone adds a new component, there is nothing additional to update. Additionally, we need to perform this in one or a few docker layers, so statically listing all components is not an option.

Unfortunately the Docker ADD/COPY support for directory wildcards doesn't end preserve the directory structure during the copy.

So the only alternative is as you suggest to COPY all of the files into a container. But this ruins build caching.

We can convert this script input into a build arg, but you are still going to need to run the script to generate the build arg.

Another approach is to statically list all components in a file and then to ensure that the build fails when a new component is added and not added to the static list.

Reply:
Why would copying the entire `.docker` directory generated in a builder container ruin the docker cache?If the checksums of all the Cargo files does not change, docker automatically detects there is no change after `COPY --from=gen-manifests /workdir/.docker ./` and uses build cache effectively

Reply:
> Why would copying the entire .docker directory generated in a builder container ruin the docker cache? 
That's not it. This script runs outside the docker context and thus has access to all files and can find all the Cargo.toml. To run it inside a builder container then you end up needing to COPY all the files into the build container, which will ruin the cache.

Reply:
Yes, COPY all the files to the *prepare* container will invalidate the cache of the *prepare* container if any file change occurs, but when COPY the collected `Cargo.yaml` files to the current *builder* container the cache is only invalidated if any `Cargo.yaml` file is changed, which is as expected.

Reply:
Yes, this works! I update the PR.